### PR TITLE
Handle forward references in a forward-compatible way

### PR DESCRIPTION
**Pull request by Teyras**

### CODE REVIEW

File: apistrap/utils.py [hunk 1]
```diff
@@ -34,3 +36,33 @@ def snake_to_camel(value):
...
+    if not isinstance(annotation, str):
+        return annotation
+
+    if globs is None:
+        globs = get_function_perspective_globals(function, level + 1, level + 1 + search_stack_depth)
+
+    return eval(annotation, globs)
+
+
+def get_type_hints(function):
```
Comment (petrbel) [inline apistrap/utils.py]: missing param/return typehints & brief doc

File: apistrap/utils.py [hunk 2]
```diff
@@ -34,3 +36,33 @@ def snake_to_camel(value):
...
+    globs = {}
+    if level != max_level:
+        stack = inspect.stack()
+        for frame in stack:
+            if frame.filename == inspect.getsourcefile(function):
+                globs.update(frame.frame.f_globals)
+    return globs
+
+
+def resolve_fw_decl(function, annotation, globs=None, level=0, search_stack_depth=2):
```
Comment (petrbel) [inline apistrap/utils.py]: missing param/return typehints & brief doc

maybe rename to `_resolve_fw_decl`?

File: apistrap/utils.py [hunk 3]
```diff
@@ -34,3 +36,33 @@ def snake_to_camel(value):
     """
     result = "".join(x.capitalize() or "_" for x in value.split("_"))
     return result[0].lower() + result[1:]
+
+
+def get_function_perspective_globals(function, level=0, max_level=None):
```
Comment (petrbel) [inline apistrap/utils.py]: missing param/return typehints & brief doc

maybe rename to `_get_function_perspective_globals` and leave without docs

File: apistrap/decorators.py [hunk 4]
```diff
@@ -201,24 +198,24 @@ def wrapper(*args, **kwargs):
...
-        _add_ignored_param(wrapper, request_arg.name)
+        _add_ignored_param(wrapper, request_arg)
         return wrapper
 
     def _check_request_type(self, *args, **kwargs):
         if self._get_request_content_type(*args, **kwargs) != "application/json":
             raise ApiClientError("Unsupported media type, JSON is expected")
 
-    def _process_request_kwargs(self, body, signature, request_arg, *args, **kwargs):
+    def _process_request_kwargs(self, body, signature, request_arg: str, *args, **kwargs):
```
Comment (petrbel) [inline apistrap/decorators.py]: missing typehints for `body` and `signature`.. or are the too complicated?

File: apistrap/utils.py [hunk 5]
```diff
@@ -34,3 +36,51 @@ def snake_to_camel(value):
...
+    """
+    if annotation is None:
+        raise ValueError("NoneType parameter annotations are not supported")
+
+    if not isinstance(annotation, str):
+        return annotation
+
+    globs = get_function_perspective_globals(function)
+
+    return eval(annotation, globs)
```
Comment (blazekadam) [inline apistrap/utils.py]: whoa

**Comment by blazekadam** [comment]
we can ignore the coverage

**Comment by Teyras** [comment]
@blazekadam Well, once (if?) Python implements looking up type annotations in `inspect`, we can remove this again.

**Pull request merged**